### ISSUE TITLE
TypeMeta is empty on .List() resources

### ISSUE THREAD

**Issue by gprossliner** [kubernetes/client-go#1328]
This is basically a reopen from #541 because I did not see why @shsjshentao closed this in the first place.

After listing / getting resources with client-go (tested with Deployments, StatefulSets and DaemonSets), `obj.TypeMeta` is empty. The fields `Kind` and `ApiVersion` are empty strings.

After the REST call and JSON deserialisation the data is still available, but is is explicitly cleared:

https://github.com/kubernetes/apimachinery/blob/02a41040d88da08de6765573ae2b1a51f424e1ca/pkg/runtime/helper.go#L255C8-L263
```go
// Decode does not do conversion. It removes the gvk during deserialization.
func (d WithoutVersionDecoder) Decode(data []byte, defaults *schema.GroupVersionKind, into Object) (Object, *schema.GroupVersionKind, error) {
	obj, gvk, err := d.Decoder.Decode(data, defaults, into)
	if obj != nil {
		kind := obj.GetObjectKind()
		// clearing the gvk is just a convention of a codec
		kind.SetGroupVersionKind(schema.GroupVersionKind{})
	}
	return obj, gvk, err
}
```
Is there any specific reason while this is necessary? This is valuable information that I need later to set `OwnerReferences` and for logging. I tried with git blame to check for the commit / PR this was introduced, but there was no information.

**Comment by dprotaso** [comment]
Does the API server return the type meta for list entries?

**Comment by gprossliner** [comment]
Yes. It is returned, but reset by client-go. After fetching from the server all information is present, but the `func (d WithoutVersionDecoder) Decode` func (see posting) clear that information.

**Comment by dprotaso** [comment]
/lifecycle frozen

Should this issue be in k/k repo?

**Comment by liggitt** [comment]
I think this is a duplicate of https://github.com/kubernetes/kubernetes/issues/3030 and https://github.com/kubernetes/kubernetes/issues/80609

**Comment by liggitt** [comment]
closing as a duplicate, see the two existing k/k issues (one tracking how now all server responses include apiVersion / kind data on individual items in lists, the other tracking how the client decoding methods clear those fields when they *are* populated in the server as part of decoding into typed go structs)
